lib/iris/tests/unit/coords/test_AuxCoord.py

```diff
@@ -0,0 +1,764 @@
+        result = coord.has_lazy_points()
+        self.assertTrue(result)
+
+    def test_lazy_core_realise(self):
+        coord = AuxCoord(self.pts_lazy)
+        coord.points
+        result = coord.has_lazy_points()
+        self.assertFalse(result)
+
+
+@skip_lazy_auxcoords
+class Test_has_lazy_bounds(tests.IrisTest):
+    def setUp(self):
+        setup_test_arrays(self)
+
+    def test_real_core(self):
+        coord = AuxCoord(self.pts_real, bounds=self.bds_real)
+        result = coord.has_lazy_bounds()
+        self.assertFalse(result)
+
+    def test_lazy_core(self):
+        coord = AuxCoord(self.pts_real, bounds=self.bds_lazy)
+        result = coord.has_lazy_bounds()
+        self.assertTrue(result)
+
+    def test_lazy_core_realise(self):
+        coord = AuxCoord(self.pts_real, bounds=self.bds_lazy)
+        coord.bounds
+        result = coord.has_lazy_bounds()
+        self.assertFalse(result)
+
+
+@skip_lazy_auxcoords
+class Test_bounds_dtype(tests.IrisTest):
+    def test_i16(self):
+        test_dtype = np.int16
+        coord = AuxCoord([1], bounds=np.array([[0, 4]], dtype=test_dtype))
+        result = coord.bounds_dtype
+        self.assertTrue(result == test_dtype)
+
+    def test_u16(self):
+        test_dtype = np.uint16
+        coord = AuxCoord([1], bounds=np.array([[0, 4]], dtype=test_dtype))
+        result = coord.bounds_dtype
+        self.assertTrue(result == test_dtype)
+
+    def test_f16(self):
+        test_dtype = np.float16
+        coord = AuxCoord([1], bounds=np.array([[0, 4]], dtype=test_dtype))
+        result = coord.bounds_dtype
+        self.assertTrue(result == test_dtype)
+
+
+def lazyness_string(data):
+    return 'lazy' if isinstance(data, da.Array) else 'real'
+
+
+_DEBUG_ALL_TYPES = _DEBUG
+
+
+def coords_all_dtypes_and_lazynesses(self, dtypes=(np.float64, np.int16)):
+    # Generate coords with all possible types of points and bounds, and all
+    # of the given dtypes.
+    points_types = ['real']
+    bounds_types = ['no', 'real']
+    if _AUXCOORD_LAZY_API:
+        points_types.append('lazy')
+        bounds_types.append('lazy')
+    for dtype in (np.float64, np.int16):
+        for points_type_name in points_types:
+            for bounds_type_name in bounds_types:
+                pts = np.asarray(self.pts_real, dtype=dtype)
+                bds = np.asarray(self.bds_real, dtype=dtype)
+                if points_type_name == 'lazy':
+                    pts = da.from_array(pts, pts.shape)
+                if bounds_type_name == 'lazy':
+                    bds = da.from_array(bds, bds.shape)
+                elif bounds_type_name == 'no':
+                    bds = None
+                coord = AuxCoord(pts, bounds=bds)
+                result = (coord, points_type_name, bounds_type_name)
+                if _DEBUG_ALL_TYPES:
+                    msg = '\nTest result type {}({} {})'
+                    print(msg.format(coord.dtype, result[1], result[2]))
+                yield result
```

Comment:
Although I guess if you went the `itertools` route then this bit would not work so nicely...